[Ng Chee Kiang] iP

CheeKiangg opened the pull request.

CheeKiangg committed:
Change Duke to CodeCatalyst

CheeKiangg committed:
Add level-1

echo commands from user

CheeKiangg committed:
Add Level-2

store text and display back to user when requested

CheeKiangg committed:
Add Level-3

task can be mark as done or unmark as not done

CheeKiangg committed:
Improve readability and error handling

CheeKiangg committed:
Add Level-4

add support for tracking 3 types of tasks: Todo, Deadline, Event

CheeKiangg committed:
Add A-TextUiTesting

CheeKiangg committed:
Add A-CodeQuality

improve code quality

CheeKiangg committed:
Refactor handleMarkCommand and handleUnmarkCommand into modular methods.

- Extracted task number parsing into a separate method `extractTaskNumber()`.
- Moved task status updating logic into a new method `changeTaskStatus()`.
- Updated `handleTaskStatusChange()` to orchestrate the task marking and unmarking process.
- Improved code readability, maintainability, and adherence to the Single Responsibility Principle.

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+    private static final int MAX_TASKS = 100;
+
+    public static void main(String[] args) {
+        Scanner scanner = new Scanner(System.in);
+
+        Task[] tasks = new Task[MAX_TASKS];
+        int taskCount = 0;
+
```
The set up of scanner, tasks, and taskCount could be refactored out into a separate set up function.

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+        Task[] tasks = new Task[MAX_TASKS];
+        int taskCount = 0;
+
+        printGreeting();
+
+        while (true) {
+            String input = scanner.nextLine();
+            printDivider();
+
```
Main logic and flow for the chatbot could be refactored out into a separate loop function.

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+            String input = scanner.nextLine();
+            printDivider();
+
+            if (input.equals("bye")) {
+                printGoodbye();
+                break;
+            } else if (input.equals("list")) {
+                printTaskList(tasks, taskCount);
+            } else if (input.startsWith("mark ")) {
+                handleTaskStatusChange(tasks, taskCount, input, true);
+            } else if (input.startsWith("unmark ")) {
+                handleTaskStatusChange(tasks, taskCount, input, false);
+            } else if (input.startsWith("todo ")) {
+                taskCount = addTask(tasks, taskCount, new Todo(input.substring(5)));
+            } else if (input.startsWith("deadline ")) {
+                taskCount = addDeadlineTask(tasks, taskCount, input);
+            } else if (input.startsWith("event ")) {
+                taskCount = addEventTask(tasks, taskCount, input);
+            } else {
+                taskCount = addTask(tasks, taskCount, new Task(input));
```
Using switch-case statements would be more readable and efficient.

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    private static void printGreeting() {
+        printDivider();
```
Could have a separate function to encapsulate  text within dividers, as opposed to printing dividers twice.

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+        Task[] tasks = new Task[MAX_TASKS];
+        int taskCount = 0;
+
+        printGreeting();
```
Good job abstracting out the greeting into a function!

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+    private static void handleTaskStatusChange(Task[] tasks, int taskCount, String input, boolean isMark) {
+        int taskNumber = extractTaskNumber(input, isMark);
+        if (taskNumber == -1) {
+            return;  // Invalid task number, already handled in extractTaskNumber.
```
Good comment to help understand the rationale for an early return!

Dri-water commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+     */
+    private static void changeTaskStatus(Task[] tasks, int taskCount, int taskNumber, boolean isMark) {
+        int taskIndex = taskNumber - 1;
+        if (taskIndex >= 0 && taskIndex < taskCount) {
+            if (isMark) {
+                tasks[taskIndex].markAsDone();
+                System.out.println("         Nice! I've marked this task as done:");
+            } else {
+                tasks[taskIndex].markAsNotDone();
+                System.out.println("         Ok, I've marked this task as not done yet:");
+            }
+            System.out.println("         " + tasks[taskIndex]);
+        } else {
+            System.out.println("         Invalid task number.");
+        }
+    }
```
The use of so many if-else statements is a bit confusing. Could be refactored.

Dri-water reviewed (commented):
Good job overall!

ehz0ah commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+
+        printGreeting();
+
+        while (true) {
```
Good job applying the SLAP principle here!

ehz0ah commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+    }
+
+    private static void printDivider() {
+        System.out.println("        ________________________________________________________\n");
```
Consider refactoring the divider string into a constant instead.
```suggestion
        private static final String SEPARATOR = "        _____________________________________\n";
        System.out.println(SEPARATOR);
```

ehz0ah commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+     * @return The valid task number, or -1 if invalid.
+     */
+    private static int extractTaskNumber(String input, boolean isMark) {
+        int prefixLength = isMark ? 5 : 7; // length of "mark " or "unmark "
```
To reduce magic numbers, consider replacing 5 and 7 with defined constants.

ehz0ah commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+    /**
+     * Change the status of the task to done or not done.
+     *
+     * @param tasks The array of tasks.
```
I like that you included some Javadoc comments. Keep it up!

ehz0ah commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,130 @@
+import java.util.Scanner;
+
+public class CodeCatalyst {
```
Consider refactoring some code into a new class as this class seems to have multiple purposes.

ehz0ah reviewed (commented):
Good job on the use of OOP concepts and code quality. Just some nits to fix and you're good to go. Keep it up!

CheeKiangg committed:
Refactor Task class to abstract and update CodeCatalyst for compatibility

- Made Task an abstract class and updated subclasses (Todo, Event, Deadline) to implement toString method appropriately.
- Removed direct instantiation of Task in CodeCatalyst.

CheeKiangg committed:
Refactor input handling to use switch-case for improved readability.

- Replaced if-else statements with a switch-case structure for handling user commands.
- Preserved functionality for all valid commands (bye, list, mark, unmark, todo, deadline, event).

CheeKiangg committed:
Merge branch 'branch-Level-5'

CheeKiangg committed:
refactor: extract input processing and error handling from main loop

- Moved input processing logic and error handling into a separate `processInput()` method.
- Reduced indentation in the main loop by simplifying the `while(true)` structure.
- Improved readability by isolating the `try-catch` block to handle errors outside of the main loop.

CheeKiangg committed:
refactor: organize files into packages and update runtest.bat

- Created new packages to organize the project structure (e.g., task package and codecatalyst package).
- Moved relevant classes into their respective packages to improve maintainability and clarity.
- Updated `runtest.bat` to reflect the new package structure, ensuring proper compilation and execution.
- Ensured that all imports are updated to align with the new package hierarchy.

CheeKiangg committed:
Merge branch 'branch-A-Packages'

CheeKiangg committed:
refactor: switch from fixed-size Task[] array to ArrayList<Task> for dynamic task management

- Replaced the fixed-size `Task[]` array with `ArrayList<Task>` to enable dynamic task management.
- Removed the `taskCount` variable as the size of the task list is now managed by `ArrayList`.
- Updated all relevant methods in `CodeCatalyst` to work with `ArrayList<Task>` instead of `Task[]`.

CheeKiangg committed:
feat: add task deletion feature

- Implemented the `delete` command to allow users to remove tasks by specifying the task number.
- Handled invalid task indices by throwing `CodeCatalystException` with appropriate error messages.
- Updated task indexing to use 1-based indexing for user interaction, while maintaining 0-based indexing internally.

Kobot7 commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,205 @@
+                break;
+            default:
+                throw new CodeCatalystException("Invalid input! Please enter a valid command.");
+                // break;
```
Delete commented code

Kobot7 commented on src/main/java/CodeCatalyst.java:
```diff
@@ -0,0 +1,205 @@
+                throw new CodeCatalystException("Invalid input! Please enter a valid command.");
+                // break;
+            }
+        } catch (CodeCatalystException e) {
```
Is the variable name "e" informative enough?

Kobot7 commented on src/main/java/codecatalyst/CodeCatalyst.java:
```diff
@@ -0,0 +1,212 @@
+     * @return The valid task number, or -1 if invalid.
+     */
+    private static int extractTaskNumber(String input, boolean isMark) {
+        int prefixLength = isMark ? 5 : 7; // length of "mark " or "unmark "
```
Consider using named constants for the magic numbers to make it more intuitive. Good call for commenting to explain what it is though!

Kobot7 commented on src/main/java/codecatalyst/task/Todo.java:
```diff
@@ -0,0 +1,23 @@
+    /**
+     * Constructs a Todo task with the specified description
+     *
+     * @param description The task description
```
Ensure that there is punctuation behind each parameter description

Kobot7 commented on src/main/java/codecatalyst/CodeCatalyst.java:
```diff
@@ -0,0 +1,212 @@
+            return Integer.parseInt(input.substring(prefixLength));
+        } catch (NumberFormatException invalidNumberFormat) {
+            System.out.println("         " + input.substring(prefixLength) + " is not a number");
+            return -1;
```
Avoid using magic numbers

CheeKiangg committed:
feat: add saveTasksToFile functionality to persist tasks

- Implemented `saveTasksToFile` to save tasks to the file.
- Tasks are saved in the specified format for `Todo`, `Deadline`, and `Event` tasks.
- The file is updated whenever a task is added, marked, or unmarked.

CheeKiangg committed:
Merge branch 'branch-Level-6'

CheeKiangg committed:
Merge branch 'branch-Level-7'

CheeKiangg committed:
feat: add functionality to save task deletions to file

- Implemented the feature to persist task deletions in text file.
- After a task is deleted, the task list is saved to the file to reflect the updated state.

CheeKiangg committed:
Refactor: Extract saveTasksWithHandling method, create specific exceptions, and eliminate magic strings/numbers

- Extracted repeated task-saving logic into a new `saveTasksWithHandling()` method to follow SRP and reduce code duplication.

CheeKiangg committed:
feat: Add Ui class to handle user interactions

- Extracted user input and output handling into a separate Ui class.
- Methods added for reading commands, displaying task lists, showing error messages, and displaying welcome and goodbye messages.
- This refactor improves separation of concerns by isolating user interaction logic from the main application logic.

CheeKiangg committed:
feat: Add Parser class to handle command parsing

- Introduced Parser class to encapsulate command parsing logic.
- Added support for parsing commands like "todo", "deadline", "event", "delete", "mark", "unmark", and "bye".
- This refactor follows SRP by isolating command parsing from the main application logic.

CheeKiangg committed:
feat: Add TaskList class to manage task operations

- TaskList class introduced to manage the collection of tasks.
- Includes methods for adding, removing, retrieving, and marking tasks as done/not done.
- This refactor improves maintainability by encapsulating task-related operations in a dedicated class.

CheeKiangg committed:
feat: Add Command class and respective commands

- Created an abstract Command class to encapsulate the execution of user commands.
- Added specific commands including AddTodoCommand, AddDeadlineCommand, AddEventCommand, DeleteCommand, MarkCommand, UnmarkCommand, ListCommand, and ExitCommand.
- Each command is responsible for executing task-related operations and interacting with the task list, storage, and UI.

CheeKiangg committed:
feat: Add Storage class for task file operations

- Extracted file handling logic into a dedicated Storage class.
- The Storage class is responsible for loading tasks from a file and saving tasks to a file.

CheeKiangg committed:
refactor: Reformat CodeCatalyst class for improved modularity

- Refactored CodeCatalyst class to use Ui, TaskList, Parser, and Storage classes.
- Reorganized main logic to delegate tasks like user input/output, command parsing, task management, and file operations to the appropriate classes.

CheeKiangg committed:
fix: Update UNMARK_LENGTH constant from 8 to 7 in Parser class

- Adjusted the UNMARK_SYMBOL constant to the correct value (7) to fix indexing issue.

CheeKiangg committed:
Merge branch 'branch-Level-8'

CheeKiangg committed:
feat: Add FindCommand class to handle search functionality

- Implemented the FindCommand class to allow searching for tasks based on a keyword.
- Added logic to find tasks in the task list whose descriptions contain the search keyword.

CheeKiangg committed:
feat: Update Parser class to support find command

- Added logic to parse the `find` command.
- Passes the search keyword to the FindCommand for execution.

CheeKiangg committed:
feat: Add methods to Ui class for displaying search results

- Added printMatchingTasks method to display tasks that match the search keyword.
- Added printNoMatchingTask method to inform users when no matching tasks are found.

CheeKiangg committed:
Merge pull request #1 from CheeKiangg/branch-Level-9

Branch level 9

CheeKiangg committed:
feat: Add comprehensive JavaDoc comments for most classes and methods

- Ensured that the comments follow best practices for professional software documentation.

CheeKiangg committed:
Merge branch 'branch-A-JavaDoc'

CheeKiangg committed:
refactor: refine InvalidTaskNumberException and improve validation in DeleteCommand, MarkCommand and UnMarkCommand.

- Updated the Commands to handle both numeric and non-numeric input errors more robustly.
- Overloaded constructor of InvalidTaskNumberException to handle different types of invalid task input.

CheeKiangg committed:
refactor: Rename InvalidTaskDescriptionException to InvalidTaskDetailException

- better reflect its broader use case as it could be the Task description, start date or end date that is missing.

CheeKiangg committed:
refactor: Improve exception handling for task-related commands

CheeKiangg committed:
docs: Add comprehensive README for CodeCatalyst

- Created a complete README file for CodeCatalyst with a detailed user guide.
- Included sections for Introduction, Quick Start, Features, FAQ, and Command Summary.
- Provided instructions on how to add, list, mark, unmark, delete, and find tasks.
- Added information on automatic saving of tasks and the file format used to store tasks.
- Ensured the README is structured to guide users through all functionalities with clear examples and command usage.

CheeKiangg committed:
style: Minor formatting adjustments

- Refined spacing, indentation, and alignment for consistency.
- Fixed minor typos and adjusted layout in comments and documentation sections.

CheeKiangg committed:
docs: Add comprehensive README in doc folder for CodeCatalyst

- Created a complete README file for CodeCatalyst with a detailed user guide.
- Included sections for Introduction, Quick Start, Features, FAQ, and Command Summary.
- Provided instructions on how to add, list, mark, unmark, delete, and find tasks.
- Added information on automatic saving of tasks and the file format used to store tasks.
- Ensured the README is structured to guide users through all functionalities with clear examples and command usage.

CheeKiangg committed:
Refactor: Update required class members to final and specify exception types in command classes

- Updated class member variables to `final` where applicable to ensure immutability and improve clarity.
- Refined exception handling in command classes by specifying the exact exceptions thrown:
  - Replaced generic `throws Exception` with more specific exception types like `InvalidTaskNumberException` and `IOException` in `Command` subclasses.

CheeKiangg committed:
Refactor: Extract task-to-string conversion logic into a separate method in Storage class

- Created `convertTaskToLine` method to encapsulate the logic for converting a task to its string representation for file storage.
- Improved code readability and maintainability by removing duplicated code for handling `Todo`, `Deadline`, and `Event` tasks.
- Reduced icomplexity and adhering to the Single Responsibility Principle.

CheeKiangg committed:
Update: Modify runtest.bat and runtest.sh to compile and test CodeCatalyst with enhanced input handling

CheeKiangg committed:
Refactor: minor code cleanup and improvements

CheeKiangg committed:
Test: update text-ui file with multiple test inputs for functionality verification

CheeKiangg committed:
Docs: update README with link to the JAR file

CheeKiangg committed:
update: removed file

CheeKiangg committed:
update: delete files

Ensure that github reflects the correct files in the most updated program.
Note that there are no edits to any code, just to sync with upstream repo
